src/core/mutation.ts

```diff
@@ -190,7 +190,6 @@ export class Mutation<
         )
       )
       .then(() => {
-        this.dispatch({ type: 'success', data })
```

Comment:
one problem I'm seeing with this change is that with mutations, we are intentionally `awaiting` the Promises returned from `onSuccess`, `onError` and `onSettled`, and only after that, we go dispatch success / error state. With your change, I think we now go into that state too early. It could also be that the promise returned from `onSuccess` fails, which should put the query into `error` state, but that's not the case now anymore.

Please also have a look at the `alpha` branch, where I've refactored this code to async/await:

https://github.com/tannerlinsley/react-query/blob/f42689951de3ef5c56150bcc90fffb0cdd54bbd5/src/core/mutation.ts#L190-L263

Reply:
The all the then chaining gave me a hard time not going to lie 😢 

I can revert that and leave the dispatch for success on line [193](https://github.com/tannerlinsley/react-query/blob/master/src/core/mutation.ts#L193).

The root issue of this is the error part. The error gets dispatched on line [225](https://github.com/tannerlinsley/react-query/blob/master/src/core/mutation.ts#L225) but because it is not thrown and the promise is not rejected yet it puts as in a unknown state (user defaults such as useErrorBoundary is not trigger)

Give me just a moment I will clip all this in the description.

Reply:
@TkDodo undone the change :D but still removed the dispatch at the [throw](https://github.com/tannerlinsley/react-query/blob/master/src/core/mutation.ts#L225) as I don't think that holds any value since when it is rejected dispatch gets called again. 

Let me know if that is incorrect / needs to be changed.